<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#Приводим-названия-столбцов-к-нижнему-регистру" data-toc-modified-id="Приводим-названия-столбцов-к-нижнему-регистру-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Приводим названия столбцов к нижнему регистру</a></span></li><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Проверка-на-дубликаты" data-toc-modified-id="Проверка-на-дубликаты-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Проверка на дубликаты</a></span></li><li><span><a href="#Удаление-лишних-столбцов" data-toc-modified-id="Удаление-лишних-столбцов-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Удаление лишних столбцов</a></span></li><li><span><a href="#Преобразование-категориальных-признаков-в-численные-(OHE)" data-toc-modified-id="Преобразование-категориальных-признаков-в-численные-(OHE)-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Преобразование категориальных признаков в численные (OHE)</a></span></li><li><span><a href="#Выделение-целевого-признака" data-toc-modified-id="Выделение-целевого-признака-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Выделение целевого признака</a></span></li><li><span><a href="#Масштабирование-численных-признаков" data-toc-modified-id="Масштабирование-численных-признаков-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Масштабирование численных признаков</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Исследование-баланса-классов" data-toc-modified-id="Исследование-баланса-классов-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Исследование баланса классов</a></span></li><li><span><a href="#Обучение-базовых-моделей-без-учета-дисбаланса-классов" data-toc-modified-id="Обучение-базовых-моделей-без-учета-дисбаланса-классов-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обучение базовых моделей без учета дисбаланса классов</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Upsampling" data-toc-modified-id="Upsampling-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Upsampling</a></span></li><li><span><a href="#Downsampling" data-toc-modified-id="Downsampling-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Downsampling</a></span></li><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Взвешивание классов</a></span></li><li><span><a href="#Оптимизация-гиперпараметров-случайного-леса" data-toc-modified-id="Оптимизация-гиперпараметров-случайного-леса-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Оптимизация гиперпараметров случайного леса</a></span></li><li><span><a href="#Случайный-лес-с-оптимальными-гиперпараметрами-и-взвешиванием-классов" data-toc-modified-id="Случайный-лес-с-оптимальными-гиперпараметрами-и-взвешиванием-классов-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Случайный лес с оптимальными гиперпараметрами и взвешиванием классов</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. У нас в распоряжении исторические данные о поведении клиентов и расторжении договоров с банком. 

Задача: построить модель с предельно большим значением *F1*-меры. Минимальный порог метрики - 0.59.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

**Цель исследования**: построить модель машинного обучения, которая сможет предсказать, уйдёт клиент из "Бета-Банка" в ближайшее время или нет, на основе исторических данных о поведении клиентов и расторжении договоров с банком. Модель должна достичь значения F1-меры не менее 0.59. Дополнительно будем измерять AUC-ROC и сравнивать её значение с F1-мерой.

**Ход исследования:**

1. Загрузка и подготовка данных
- Импортируем необходимые библиотеки
- Загрузим предоставленные данные
- Изучим данные: проверим на наличие пропусков и дубликатов, а также осуществим преобразования при необходимости
- Разделим данные на обучающую, валидационную и тестовую выборки
- Исследование баланса классов и обучение модели без учёта дисбаланса

2. Исследуем баланс классов в обучающей выборке
- Обучим базовые модели без учета дисбаланса классов (например, логистическая регрессия, случайный лес, градиентный бустинг)
- Оценим качество моделей с помощью метрик F1-меры и AUC-ROC
- Улучшение качества модели, учитывая дисбаланс классов

3. Применим техники борьбы с дисбалансом классов, такие как взвешивание классов, аугментация данных или андерсэмплинг/оверсэмплинг
- Обучим различные модели с учетом дисбаланса классов
- Оценим качество моделей с помощью метрик F1-меры и AUC-ROC
- Выберем лучшую модель и оптимизируем гиперпараметры при необходимости

4. Финальное тестирование
- Проведем финальное тестирование выбранной модели на тестовой выборке
- Оценим качество модели с помощью метрик F1-меры и AUC-ROC
- Сделаем выводы о результатах и выполнении целей проекта

## Подготовка данных

### Изучение данных

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import re
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, make_scorer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
# import warnings
# warnings.filterwarnings("ignore")

In [2]:
try:
    data = pd.read_csv('/datasets/Churn.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/Churn.csv')

In [3]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [5]:
data.isna().sum()

RowNumber            0
CustomerId           0
Surname              0
CreditScore          0
Geography            0
Gender               0
Age                  0
Tenure             909
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary      0
Exited               0
dtype: int64

In [6]:
data.corr(numeric_only=True)

RowNumber  CustomerId  CreditScore       Age    Tenure  \
RowNumber         1.000000    0.004202     0.005840  0.000783 -0.007322   
CustomerId        0.004202    1.000000     0.005308  0.009497 -0.021418   
CreditScore       0.005840    0.005308     1.000000 -0.003965 -0.000062   
Age               0.000783    0.009497    -0.003965  1.000000 -0.013134   
Tenure           -0.007322   -0.021418    -0.000062 -0.013134  1.000000   
Balance          -0.009067   -0.012419     0.006268  0.028308 -0.007911   
NumOfProducts     0.007246    0.016972     0.012238 -0.030680  0.011979   
HasCrCard         0.000599   -0.014025    -0.005458 -0.011721  0.027232   
IsActiveMember    0.012044    0.001665     0.025651  0.085472 -0.032178   
EstimatedSalary  -0.005988    0.015271    -0.001384 -0.007201  0.010520   
Exited           -0.016571   -0.006248    -0.027094  0.285323 -0.016761   

                  Balance  NumOfProducts  HasCrCard  IsActiveMember  \
RowNumber       -0.009067       0.007246   0.000599        0.012044   
CustomerId      -0.012419       0.016972  -0.014025        0.001665   
CreditScore      0.006268       0.012238  -0.005458        0.025651   
Age              0.028308      -0.030680  -0.011721        0.085472   
Tenure          -0.007911       0.011979   0.027232       -0.032178   
Balance          1.000000      -0.304180  -0.014858       -0.010084   
NumOfProducts   -0.304180       1.000000   0.003183        0.009612   
HasCrCard       -0.014858       0.003183   1.000000       -0.011866   
IsActiveMember  -0.010084       0.009612  -0.011866        1.000000   
EstimatedSalary  0.012797       0.014204  -0.009933       -0.011421   
Exited           0.118533      -0.047820  -0.007138       -0.156128   

                 EstimatedSalary    Exited  
RowNumber              -0.005988 -0.016571  
CustomerId              0.015271 -0.006248  
CreditScore            -0.001384 -0.027094  
Age                    -0.007201  0.285323  
Tenure                  0.010520 -0.016761  
Balance                 0.012797  0.118533  
NumOfProducts           0.014204 -0.047820  
HasCrCard              -0.009933 -0.007138  
IsActiveMember         -0.011421 -0.156128  
EstimatedSalary         1.000000  0.012097  
Exited                  0.012097  1.000000

### Приводим названия столбцов к нижнему регистру

In [7]:
def camel_to_snake(name):
    name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', name).lower()

new_columns = {col: camel_to_snake(col) for col in data.columns}
data = data.rename(columns=new_columns)
data.head()

row_number  customer_id   surname  credit_score geography  gender  age  \
0           1     15634602  Hargrave           619    France  Female   42   
1           2     15647311      Hill           608     Spain  Female   41   
2           3     15619304      Onio           502    France  Female   42   
3           4     15701354      Boni           699    France  Female   39   
4           5     15737888  Mitchell           850     Spain  Female   43   

   tenure    balance  num_of_products  has_cr_card  is_active_member  \
0     2.0       0.00                1            1                 1   
1     1.0   83807.86                1            0                 1   
2     8.0  159660.80                3            1                 0   
3     1.0       0.00                2            0                 0   
4     2.0  125510.82                1            1                 1   

   estimated_salary  exited  
0         101348.88       1  
1         112542.58       0  
2         113931.57       1  
3          93826.63       0  
4          79084.10       0

### Удаление пропусков

In [8]:
data.isnull().sum()

row_number            0
customer_id           0
surname               0
credit_score          0
geography             0
gender                0
age                   0
tenure              909
balance               0
num_of_products       0
has_cr_card           0
is_active_member      0
estimated_salary      0
exited                0
dtype: int64

In [9]:
data = data.dropna()
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   row_number        9091 non-null   int64  
 1   customer_id       9091 non-null   int64  
 2   surname           9091 non-null   object 
 3   credit_score      9091 non-null   int64  
 4   geography         9091 non-null   object 
 5   gender            9091 non-null   object 
 6   age               9091 non-null   int64  
 7   tenure            9091 non-null   float64
 8   balance           9091 non-null   float64
 9   num_of_products   9091 non-null   int64  
 10  has_cr_card       9091 non-null   int64  
 11  is_active_member  9091 non-null   int64  
 12  estimated_salary  9091 non-null   float64
 13  exited            9091 non-null   int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.0+ MB


Мы не можем восстановить пропущенные данные в колонке Tenure, и для будущего обучения моделей лучше не заполнять эти пропуски медианой, поэтому удалим пропуски.

### Проверка на дубликаты

In [10]:
data.duplicated().sum()

0

### Удаление лишних столбцов

In [11]:
data.drop(["row_number", "customer_id", "surname"], axis=1, inplace=True)
data.isnull().sum()

credit_score        0
geography           0
gender              0
age                 0
tenure              0
balance             0
num_of_products     0
has_cr_card         0
is_active_member    0
estimated_salary    0
exited              0
dtype: int64

Столбцы "row_number", "customer_id" и "surname" являются идентификаторами и не несут никакой полезной информации для моделирования оттока клиентов из банка.

Столбец "row_number" содержит просто порядковый номер строки в наборе данных, а столбец "customer_id" является уникальным идентификатором клиента. Оба этих столбца не содержат никакой информации о клиенте, которая могла бы помочь в прогнозировании его оттока из банка.

Столбец "surname" содержит информацию о фамилии клиента. Хотя фамилия может быть полезна для идентификации клиентов в реальной жизни, в контексте прогнозирования оттока клиентов из банка эта информация не является существенной и может быть удалена без потери полезной информации для модели.

### Преобразование категориальных признаков в численные (OHE)

In [12]:
encoder = OneHotEncoder(sparse_output=False, drop='first')
categorical_columns = ['geography', 'gender']

# Применение OneHotEncoder к категориальным колонкам
encoded_data = encoder.fit_transform(data[categorical_columns])
encoded_feature_names = encoder.get_feature_names_out(categorical_columns)

# Создание нового DataFrame с преобразованными колонками
encoded_df = pd.DataFrame(encoded_data, columns=encoded_feature_names)

# Удаление исходных категориальных колонок
data = data.drop(columns=categorical_columns)

# Сброс индексов исходного DataFrame
data.reset_index(drop=True, inplace=True)

# Объединение исходного DataFrame с encoded_df
data = pd.concat([data, encoded_df], axis=1)

### Выделение целевого признака

In [13]:
# Разделение данных на обучающую, валидационную и тестовую выборки (60:20:20)
train_data, temp_data = train_test_split(data, test_size=0.4, random_state=12345)
valid_data, test_data = train_test_split(temp_data, test_size=0.5, random_state=12345)

X_train = train_data.drop("exited", axis=1)
y_train = train_data["exited"]

X_valid = valid_data.drop("exited", axis=1)
y_valid = valid_data["exited"]

X_test = test_data.drop("exited", axis=1)
y_test = test_data["exited"]
print(X_train.shape)
print(X_valid.shape)
print(X_test.shape)

(5454, 11)
(1818, 11)
(1819, 11)


### Масштабирование численных признаков

In [14]:
numeric_columns = ["credit_score", "age", "tenure", "balance", "num_of_products", "estimated_salary"]

scaler = StandardScaler()
scaler.fit(X_train[numeric_columns])
X_train[numeric_columns] = scaler.transform(X_train[numeric_columns])
X_valid[numeric_columns] = scaler.transform(X_valid[numeric_columns])
X_test[numeric_columns] = scaler.transform(X_test[numeric_columns])

## Исследование задачи

### Исследование баланса классов

In [15]:
class_count = data["exited"].value_counts(normalize=True)
print("Баланс классов:\n", class_count)

Баланс классов:
 0    0.796062
1    0.203938
Name: exited, dtype: float64


### Обучение базовых моделей без учета дисбаланса классов

In [16]:
# Логистическая регрессия
logreg = LogisticRegression(random_state=12345)
logreg.fit(X_train, y_train)
logreg_valid_pred = logreg.predict(X_valid)

logreg_f1 = f1_score(y_valid, logreg_valid_pred)
logreg_proba = logreg.predict_proba(X_valid)
logreg_proba = logreg_proba[:, 1]

logreg_roc_auc = roc_auc_score(y_valid, logreg_proba)

In [17]:
# Случайный лес
rfc = RandomForestClassifier(random_state=12345)
rfc.fit(X_train, y_train)
rfc_valid_pred = rfc.predict(X_valid)

rfc_f1 = f1_score(y_valid, rfc_valid_pred)
rfc_proba = rfc.predict_proba(X_valid)
rfc_proba = rfc_proba[:, 1]

rfc_roc_auc = roc_auc_score(y_valid, rfc_proba)

In [18]:
# Вывод результатов
print("\nМетрики базовых моделей без учета дисбаланса классов:")
print("Логистическая регрессия: F1 = {:.3f}, AUC-ROC = {:.3f}".format(logreg_f1, logreg_roc_auc))
print("Случайный лес: F1 = {:.3f}, AUC-ROC = {:.3f}".format(rfc_f1, rfc_roc_auc))


Метрики базовых моделей без учета дисбаланса классов:
Логистическая регрессия: F1 = 0.304, AUC-ROC = 0.774
Случайный лес: F1 = 0.590, AUC-ROC = 0.863


### Вывод

- Мы исследовали баланс классов и обучили базовые модели без учета дисбаланса классов.  
- Результаты показали, что существует дисбаланс классов, который может снижать качество моделей. Он составляет 80/20. 
- Мы выяснили, что при использовании базовых моделей без учета дисбаланса классов, метрики F1 и AUC-ROC оказались недостаточными для достижения целевого значения F1-меры.

Обучение базовых моделей без учета дисбаланса классов позволило нам сравнить их производительность и выбрать случайный лес в качестве отправной точки для дальнейшей оптимизации модели с учетом дисбаланса классов.

## Борьба с дисбалансом

### Upsampling

In [19]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

X_upsampled_train, y_upsampled_train = upsample(X_train, y_train, repeat=4)

In [20]:
rfc_upsampling = RandomForestClassifier(random_state=12345, n_jobs=-1)
rfc_upsampling.fit(X_upsampled_train, y_upsampled_train)
rfc_upsampling_pred = rfc_upsampling.predict(X_valid)
rfc_upsampling_f1 = f1_score(y_valid, rfc_upsampling_pred)
rfc_upsampling_proba = rfc_upsampling.predict_proba(X_valid)
rfc_upsampling_proba = rfc_upsampling_proba[:, 1]

rfc_upsampling_roc_auc = roc_auc_score(y_valid, rfc_upsampling_proba)

print("Случайный лес: F1 = {:.3f}, AUC-ROC = {:.3f}".format(rfc_upsampling_f1, rfc_upsampling_roc_auc))

Случайный лес: F1 = 0.609, AUC-ROC = 0.859


### Downsampling

In [21]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

X_downsampled_train, y_downsampled_train = downsample(X_train, y_train, fraction=0.25)

In [22]:
rfc_downsampling = RandomForestClassifier(random_state=12345, n_jobs=-1)
rfc_downsampling.fit(X_downsampled_train, y_downsampled_train)
rfc_downsampling_pred = rfc_downsampling.predict(X_valid)
rfc_downsampling_f1 = f1_score(y_valid, rfc_downsampling_pred)
rfc_downsampling_proba = rfc_downsampling.predict_proba(X_valid)
rfc_downsampling_proba = rfc_downsampling_proba[:, 1]

rfc_downsampling_roc_auc = roc_auc_score(y_valid, rfc_downsampling_proba)

print("Случайный лес: F1 = {:.3f}, AUC-ROC = {:.3f}".format(rfc_downsampling_f1, rfc_downsampling_roc_auc))

Случайный лес: F1 = 0.589, AUC-ROC = 0.864


### Взвешивание классов

In [23]:
rfc_balanced = RandomForestClassifier(class_weight='balanced', random_state=12345)
rfc_balanced.fit(X_train, y_train)
rfc_balanced_valid_pred = rfc_balanced.predict(X_valid)

rfc_balanced_f1 = f1_score(y_valid, rfc_balanced_valid_pred)
rfc_balanced_proba = rfc_balanced.predict_proba(X_valid)
rfc_balanced_proba = rfc_balanced_proba[:, 1]

rfc_balanced_roc_auc = roc_auc_score(y_valid, rfc_balanced_proba)

print("Случайный лес: F1 = {:.3f}, AUC-ROC = {:.3f}".format(rfc_balanced_f1, rfc_balanced_roc_auc))

Случайный лес: F1 = 0.557, AUC-ROC = 0.862


### Оптимизация гиперпараметров случайного леса

In [24]:
rfc_balanced_grid = {
    'n_estimators': [10, 50, 100],
    'max_depth': [5, 10, 15],
    'min_samples_leaf': [1, 2, 4],
    'min_samples_split': [2, 5, 10]
}

rfc_balanced_cv = GridSearchCV(
    RandomForestClassifier(class_weight='balanced', random_state=12345),
    param_grid=rfc_balanced_grid,
    scoring=make_scorer(f1_score),
    cv=5,
    verbose=1,
    n_jobs=-1
)

rfc_balanced_cv.fit(X_train, y_train)
print("\nЛучшие гиперпараметры для случайного леса с взвешиванием классов:")
print(rfc_balanced_cv.best_params_)

Fitting 5 folds for each of 81 candidates, totalling 405 fits

Лучшие гиперпараметры для случайного леса с взвешиванием классов:
{'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 10, 'n_estimators': 100}


### Случайный лес с оптимальными гиперпараметрами и взвешиванием классов

In [25]:
rfc_balanced_best = RandomForestClassifier(
    class_weight='balanced',
    random_state=12345,
    **rfc_balanced_cv.best_params_
)

rfc_balanced_best.fit(X_train, y_train)
rfc_balanced_best_valid_pred = rfc_balanced_best.predict(X_valid)

rfc_balanced_best_f1 = f1_score(y_valid, rfc_balanced_best_valid_pred)
rfc_balanced_best_proba = rfc_balanced_best.predict_proba(X_valid)
rfc_balanced_best_proba = rfc_balanced_best_proba[:, 1]

rfc_balanced_best_roc_auc = roc_auc_score(y_valid, rfc_balanced_best_proba)

# Вывод результатов
print("\nМетрики модели случайного леса с оптимальными гиперпараметрами и взвешиванием классов:")
print("F1 = {:.3f}, AUC-ROC = {:.3f}".format(rfc_balanced_best_f1, rfc_balanced_best_roc_auc))


Метрики модели случайного леса с оптимальными гиперпараметрами и взвешиванием классов:
F1 = 0.645, AUC-ROC = 0.873


### Вывод

Использование различных методов борьбы с дисбалансом классов позволило нам значительно улучшить производительность модели случайного леса. В частности, мы сравнили модель на трех методах борьбы с дисбалансом классов:
- взвешивание классов
- увеличение выборки(upsampling)
- уменьшение выборки(downsampling).

Мы выбрали оптимальные гиперпараметры на обучающей выборке с помощью поиска по сетке, затем мы оценили производительность моделей на валидационной выборке и выбрали модель с наилучшей метрикой F1.

Наилучшей моделью оказалась модель случайного леса, обученная с использованием метода взвешивания классов, оптимальными гиперпараметрами, найденными с помощью поиска по сетке. Эта модель достигла метрики F1-меры на уровне 0.65, что превосходит более простые модели, обученные без учета дисбаланса классов, или без оптимизации гиперпараметров.

## Тестирование модели

In [26]:
# Предсказания для тестовой выборки
rfc_balanced_best_test_pred = rfc_balanced_best.predict(X_test)

# Расчет метрик на тестовой выборке
rfc_balanced_best_test_f1 = f1_score(y_test, rfc_balanced_best_test_pred)
rfc_balanced_best_test_proba = rfc_balanced_best.predict_proba(X_test)
rfc_balanced_best_test_proba = rfc_balanced_best_test_proba[:, 1]

rfc_balanced_best_test_roc_auc = roc_auc_score(y_test, rfc_balanced_best_test_proba)

# Вывод результатов
print("\nМетрики лучшей модели (случайный лес) на тестовой выборке после дополнительного обучения на валидационной выборке:")
print("F1 = {:.3f}, AUC-ROC = {:.3f}".format(rfc_balanced_best_test_f1, rfc_balanced_best_test_roc_auc))


Метрики лучшей модели (случайный лес) на тестовой выборке после дополнительного обучения на валидационной выборке:
F1 = 0.611, AUC-ROC = 0.860


## Вывод

В ходе проекта мы построили модель для прогнозирования оттока клиентов из банка. Используя данные о поведении клиентов и расторжении договоров с банком, мы построили модели на основе логистической регрессии и случайного леса, чтобы предсказать, уйдет ли клиент из банка в ближайшее время или нет.

В процессе обучения моделей мы столкнулись с проблемой дисбаланса классов, что привело к низкой производительности моделей. При обучении без учета дисбаланса классов мы увидели значительную разницу между случайным лесом и логистической регрессией, поэтому для дальнейших исследований решили использовать лучшую - случайный лес. Для борьбы с дисбалансом классов мы использовали методы взвешивания классов, upsampling и downsampling, чтобы улучшить производительность моделей.

Был использован поиск по сетке, чтобы определить оптимальные гиперпараметры для моделей и сравнили производительность на валидационной выборке.

Используя тестовую выборку, мы оценили производительность наших лучших моделей. Лучшей моделью оказалась модель случайного леса с взвешиванием классов и оптимальными гиперпараметрами, которая достигла метрики F1-меры на уровне 0.61.

Таким образом, мы разработали модель с предельно большим значением F1-меры, которая может использоваться для прогнозирования оттока клиентов из банка. Результаты проекта могут быть полезны банкам, которые хотят повысить свою эффективность и сохранить текущих клиентов.